### ISSUE TITLE
Mongo DB queries - execute $match before $project

### ISSUE THREAD

**Issue by jezzarax** [metabase/metabase#4216]
Hi,

I noticed that when using mongodb with metabase, the aggregation pipeline that is built to pull the data from mongodb contains $project element before $match. That prevents mongodb from using indexes created on the collection and reduces the performance.

Here's the example of a metabase log I got while running a question inside metabase:

```
01-23 17:53:51 DEBUG metabase.query-processor :: NATIVE FORM: 😳
{:query
 [{"$project" {"Clicks" "$Clicks", "Campaign" "$Campaign", "AdvertisingChannel" "$AdvertisingChannel", "Day" "$Day", "HourOfDay" "$HourOfDay"}}
  {"$match" {"$and" [{"Campaign" {"$eq" "XXXXX"}} {"AdvertisingChannel" {"$eq" "YYYYY"}} {"Day" {"$eq" "ZZZZZ"}}]}}
  {"$project" {"_id" "$_id", "___group" {"HourOfDay" "$HourOfDay"}, "Clicks" "$Clicks"}}
  {"$group" {"_id" "$___group", "count" {"$sum" 1}, "sum" {"$sum" "$Clicks"}}}
  {"$sort" {"_id" 1}}
  {"$project" {"_id" false, "count" true, "sum" true, "HourOfDay" "$_id.HourOfDay"}}
  {"$sort" {"HourOfDay" 1}}
  {"$limit" 25}],
 :collection "CCCCC",
 :mbql? true}

```
There's an index on Campaign and Day fields, but according to the .explain() it is never used. If I issue the same query, but with $project and $match in reversed order, the problem goes away. The problem (or actual behaviour) is described here: http://stackoverflow.com/questions/14987151/mongo-aggregate-match-before-project.
 
### Environment:
- macOs 10.12.2
- Metabase version v0.22.0 (fd04424 release-0.22.0)
- MongoDB v3.4.0

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by rblack** [comment]
Have problems with this kind of aggregations too. We have collection of about 30+ million documents and even on super simple queries matching only few docs there is huge load of db server (up to 800%) and enormous waiting time for response.

**Comment by BeHiz** [comment]
I posted a feature request to allow running of saved queries (views). Which would allow you to write your own query in native MongoDB form, thus utilizing the native indexes.

If you feel this would help, please vote for it.
https://github.com/metabase/metabase/issues/4789

**Comment by ujjwal-gupta** [comment]
Any news on this? Metabase has become completely unusable for large Mongo collections because of this.

**Comment by thomasnordquist** [comment]
I'm also experiencing severe performance issues, but instead of a very huge dataset (just 2mil), I have 20 dashboard metrics that will time out when updating the same time.
Caching is the only way I can get to a complete dashboard (by updating 3-4 times) after a few results are already cached.
Writing queries by hand yields the expected performance but is tedious and somehow defies the purpose of using metabase (and the very nice query builder).

As @jezzarax already noticed, projecting before matching enforces a full-table-scan each query.
I thought about the problem quite a lot and could not think of a case where we really need a projected field in the matching stage.

<img width="1400" alt="Bildschirmfoto 2019-03-19 um 20 26 57" src="https://user-images.githubusercontent.com/7721625/54635918-91478500-4a85-11e9-9674-0bf24f084911.png">
<img width="1400" alt="Bildschirmfoto 2019-03-19 um 20 27 37" src="https://user-images.githubusercontent.com/7721625/54635922-94427580-4a85-11e9-8768-02a12918b9ec.png">

**Comment by relferreira** [comment]
Any news about this issue?

**Comment by fabtrompet** [comment]
same problem here :(

**Comment by camsaul** [comment]
I ended up implementing this in #11832 because it was bothering me.

**Comment by fabtrompet** [comment]
> I ended up implementing this in #11832 because it was bothering me.

@camsaul thanks for the effort, I tried here and it's working on filters that I use the equality operation for example, however in the operation between, the project continues in front of the match

**Comment by salsakran** [comment]
reopening based on @fabtrompet 's feedback

**Comment by 9terz** [comment]
I also experienced this in the operation between. Any updates about the issues?

**Comment by camsaul** [comment]
Implemented by #14840
